plutus-core-spec/figures/PlutusCoreTypeCorrectness-Algorithmic-Unrestricted.tex

```diff
@@ -0,0 +1,232 @@
+        \BinaryInfC{\(\Gamma, \typeJ{\alpha}{K}\)}
+    \end{prooftree}
+
+    \begin{prooftree}
+        \AxiomC{\(\validJ{\Gamma}\)}
+        \AxiomC{$x$ is free in $\Gamma$}
+        \AxiomC{\(\hypJ{\Gamma}{\istypeJ{S}{\typeK{}}}\)}
+        \TrinaryInfC{\(\Gamma, \termJ{x}{S}\)}
+    \end{prooftree}
+
+
+    \caption{Contexts}
+    \label{fig:Plutus_core_contexts}
+\end{figure*}
+
+
+
+
+
+\begin{figure*}[t]
+    \judgmentdef{\(\hypJ{\Gamma}{\istypeJ{A}{K}}\)}{In context $\Gamma$, type $A$ has kind $K$}
+
+    \begin{prooftree}
+        \AxiomC{\(\ctxni{\Gamma}{\typeJ{\alpha}{K}}\)}
+        \RightLabel{tyvar}
+        \UnaryInfC{\(\hypJ{\Gamma}{\istypeJ{\alpha}{K}}\)}
+    \end{prooftree}
+
+    \begin{prooftree}
+        \AxiomC{\(\hypJ{\Gamma, \typeJ{\alpha}{K}}{\istypeJ{A}{\typeK{}}}\)}
+        \RightLabel{tyall}
+        \UnaryInfC{\(\hypJ{\Gamma}{\istypeJ{\allT{\alpha}{K}{A}}{\typeK{}}}\)}
+    \end{prooftree}
+
+    \begin{prooftree}
+        \AxiomC{\(\hypJ{\Gamma, \typeJ{\alpha}{\typeK{}}}{\istypeJ{A}{K}}\)}
+        \AxiomC{\(K \equiv \typeK{}\)}
+        \RightLabel{tyfix}
+        \BinaryInfC{\(\hypJ{\Gamma}{\istypeJ{\fixT{\alpha}{A}}{\typeK{}}}\)}
+    \end{prooftree}
+
+    \begin{prooftree}
+        \AxiomC{\(\hypJ{\Gamma}{\istypeJ{A}{\typeK{}}}\)}
+        \AxiomC{\(\hypJ{\Gamma}{\istypeJ{B}{\typeK{}}}\)}
+        \RightLabel{tyfun}
+        \BinaryInfC{\(\hypJ{\Gamma}{\istypeJ{\funT{A}{B }}{\typeK{}}}\)}
+    \end{prooftree}
+
+    \begin{prooftree}
+        \AxiomC{\(\hypJ{\Gamma, \typeJ{\alpha}{J}}{\istypeJ{A}{K}}\)}
+        \RightLabel{tylam}
+        \UnaryInfC{\(\hypJ{\Gamma}{\istypeJ{\lamT{\alpha}{J}{A}}{\funK{J}{K}}}\)}
+    \end{prooftree}
+
+    \begin{prooftree}
+        \AxiomC{\(\hypJ{\Gamma}{\istypeJ{A}{\funK{J}{K}}}\)}
+        \AxiomC{\(\hypJ{\Gamma}{\istypeJ{B}{J}}\)}
+        \RightLabel{tyapp}
+        \BinaryInfC{\(\hypJ{\Gamma}{\istypeJ{\appT{A}{B}}{K}}\)}
+    \end{prooftree}
+
+    \begin{prooftree}
+        \AxiomC{$tcn$ is a type constant in in Fig. \ref{fig:Plutus_core_type_constants}}
+        \AxiomC{\(\hypJ{\Gamma}{\istypeJ{A}{\sizeK{}}}\)}
+        \RightLabel{tycon}
+        \BinaryInfC{\(\hypJ{\Gamma}{\istypeJ{\conT{tcn}{A}}{\typeK{}}}\)}
+    \end{prooftree}
+
+    \caption{Kind Synthesis}
+    \label{fig:Plutus_core_kind_synthesis}
+\end{figure*}
+
+
+
+
+
+\begin{figure*}[t]
+    \judgmentdef{\(\hypJ{\Gamma}{\istermJ{M}{S}}\)}{In context $\Gamma$, term $M$ has normal type $S$}
+
+    \begin{prooftree}
+        \AxiomC{\(\ctxni{\Gamma}{\termJ{x}{S}}\)}
+        \RightLabel{var}
+        \UnaryInfC{\(\hypJ{\Gamma}{\istermJ{x}{S}}\)}
+    \end{prooftree}
+
+    \begin{prooftree}
+        \AxiomC{$cn$ has constant signature $\constsig{tcn}{s}$ in Fig. \ref{fig:Plutus_core_constants}}
+        \RightLabel{con}
+        \UnaryInfC{\(\hypJ{\Gamma}{\istermJ{cn}{\conT{tcn}{s}}}\)}
+    \end{prooftree}
+
+    \begin{prooftree}
+        \AxiomC{\(\hypJ{\Gamma, \typeJ{\alpha}{K}}{\istermJ{M}{T}}\)}
```

Comment:
Did you want to use `S` throughout for the synthesized type of the term?

Reply:
No, there's a subtle naming convention going on that comes from Phil around what types metavars are used for what sorts of things. `x` and `M` have type `A`/`T`, `y` and `N` have type `B`/`S`, etc. and when these aren't applicable I'm trying to use `C`/`R` (which is often used as the third type metavar, esp. for miscellaneous other types (eg the branches of a case)).

Reply:
Oh so there is - I got the symbols the wrong way around and I thought triple-equals was equality modulo beta. My notational instincts led me to assume assume that the triple-equals was "stronger" and hence was the one that included reduction.

Ugh, I should have another look through and make sure all the equalities are indeed the ones I expect, then.

Reply:
Whoops, this comment was supposed to be in response to the one below about `equiv`.

Reply:
Ugh, I hate metavar naming conventions. So implicit.

I've realised my actual confusion: I thought `T` wasn't in the metavar set for type values, but it is!